0003-secret-seed-storage-and-key-derivation.md

```diff
@@ -0,0 +1,195 @@
+# Secret seed storage and key derivation
+
+* Authors: Lloyd Fournier
+* Date: 2019-04-16
+
+Issue: [#672](https://github.com/comit-network/comit-rs/issues/672)
+
+## Context
+
+Currently, our node's 32 byte secret seed is stored in the config file as hex and we derive child seeds from it for each protocol execution.
+This system was invented ad-hoc and the motivation for this spike is to design a well thought out solution that is informed by other projects.
+
+I present the conclusions first and a summary of what I learned from LND.
+
+## Secret Seed Storage
+
+This section recommends how we should generate and store the secret seed by answering key questions.
+
+### Is there any way to recover anything if you have your secret seed but no database?
+
+**In practice, No.**
+
+In Theory and in the case of Ethereum only you could scan the blockchain for any HTLC looking contract paying money to an address you generated deterministically from your seed and reclaim it.
+
+In Bitcoin you can't because you'll never be able to compute the P2WSH scriptPubKey without knowing the other person's input.
+If it has already been redeemed, you could scan for redeem transactions with script that matches our Bitcoin HTLC, recover the secret and then look for the HTLC on the other chain (as above, assuming it's Ethereum).
+
+I don't think this partial recovery is worth putting the significant effort that it would require to implement.
+
+### Should it be encrypted?
+
+I think this should be optional.
+A COMIT node doesn't own the funds long term.
+It can only control the funds of protocols that are in being executed.
+This is different to LND where if someone gets access to your seed they can take all your funds regardless of whether you have channels open or not.
+
+The best attack vector against an unencrypted seed is for the attacker to get short term access to the system COMIT is running on, covertly steal their secret seed and then wait for them to execute a swap and then redeem to your own address.
+Encrypting the seed does give you protection against this.
+
+Having said that, giving an attacker access to your COMIT node will probably give them access to the HTTP API which they can use to redeem funds to their own address.
+This is probably a better attack vector than the above.
+
+note: We could prevent this attack by giving the node a way of generating the final redeem addresses without getting them from the api.
```

Comment:
```suggestion
Note: We could prevent this attack by allowing the node to generate the final redeem addresses without getting them from the api.
```

Reply:
"giving" had the meaning I wanted to convey. In order for it to do this it would have to have a way to generate it given to it from the outside.